src/translate/pddl_parser/parsing_functions.py

```diff
@@ -313,180 +654,164 @@ def parse_task(domain_pddl, task_pddl):
+            if (not check_named_block(domain_line, [":domain"]) or
+                    len(domain_line) != 2 or not isinstance(domain_line[1], str)):
+                context.error("Invalid domain name definition", domain_line,
+                              syntax=SYNTAX_TASK_DOMAIN_NAME)
+            yield domain_line[1]
+
+        requirements_opt = next(iterator)
+        assert_named_block(context, requirements_opt, [":requirements", ":objects", ":init"])
+        if requirements_opt[0] == ":requirements":
+            requirements = requirements_opt[1:]
+            objects_opt = next(iterator)
         else:
-            atom = pddl.Atom(fact[0], fact[1:])
-            check_atom_consistency(atom, initial_true, initial_false)
-            initial_true.add(atom)
-    initial.extend(initial_true)
-    yield initial
-
-    goal = next(iterator)
-    assert goal[0] == ":goal" and len(goal) == 2
-    yield parse_condition(goal[1], type_dict, predicate_dict)
-
-    use_metric = False
-    for entry in iterator:
-        if entry[0] == ":metric":
-            if entry[1] == "minimize" and entry[2][0] == "total-cost":
-                use_metric = True
-            else:
-                assert False, "Unknown metric."
-    yield use_metric
-
-    for entry in iterator:
-        assert False, entry
-
-
-def check_atom_consistency(atom, same_truth_value, other_truth_value, atom_is_true=True):
+            requirements = []
+            objects_opt = requirements_opt
+        yield parse_requirements(context, requirements)
+
+        assert_named_block(context, objects_opt, [":objects", ":init"])
+        if objects_opt[0] == ":objects":
+            with context.layer("Parsing objects"):
+                yield parse_typed_list(context, objects_opt[1:])
+            init = next(iterator)
+        else:
+            yield []
+            init = objects_opt
+
+        assert_named_block(context, init, [":init"])
+        yield parse_init(context, init)
+
+        goal = next(iterator)
+        with context.layer("Parsing goal"):
+            if (not check_named_block(goal, [":goal"]) or
+                    len(goal) != 2 or not isinstance(goal[1], list) or
+                    not goal[1]):
+                context.error("Expected non-empty goal.", syntax=SYNTAX_GOAL)
+            yield parse_condition(context, goal[1], type_dict, predicate_dict)
+
+        use_metric = False
+        for entry in iterator:
+            if isinstance(entry, list) and entry[0] == ":metric":
+                with context.layer("Parsing metric"):
+                    if len(entry) != 3 or not isinstance(entry[2], list) or len(entry[2]) != 1 or entry[1] != "minimize" or entry[2][0] != "total-cost":
+                        context.error("Invalid metric definition.", entry, syntax=SYNTAX_METRIC)
+                    use_metric = True
+        yield use_metric
+
+        for _ in iterator:
+            assert False, "This line should be unreachable"
```

Comment:
This can happen when we have malformed PDDL, i.e. when someone defines more things after metric, correct? Then we should give an error message.

As an aside, if someone does not specify a metric but other things, do they get swallowed and the PDDL accepted? I will test this.

Reply:
I can neither confirm that this can happen nor do I understand the context to meaningfully comment on this.

Reply:
I ran this and found three issues:
```
(
   ...
    (:goal ...  )
    (foo)
    (:metric minimize (total-cost))
    (bar)
)
```
Anything after `(:goal ...)` other than `(:metric ...)` is silently ignored. The assert actually never triggers because the loop above consumes all elements. But there shouldn't be a loop to ignore `(foo)` to begin with, right?

The other issue is this with this code:
```
(
   ...
    (:goal ...  )
    (:metric minimize (total-cost))
    ()
)
```
where the check `entry[0] == ":metric"` fails.